### DISCUSSION TITLE
How to transfer multiple cells to write instead of one cell at a time?

### DISCUSSION THREAD

**Discussion by ishmeets920** [cubewise-code/tm1py#761] [Q&A]
Hello All,

Hope all are doing good. 

Usecase : Need to write the data from one cube to another 

```
Current code : 
cellset = {(Dimensionlist[0], Dimensionlist[1], Dimensionlist[2], Dimensionlist[3]): Value for
            Dimensionlist[0], Dimensionlist[1], Dimensionlist[2], Dimensionlist[3], Value in df.values}
 
 tm2.cubes.cells.write_values(source_cube, cellset,skip_non_updateable=True, dimensions=Target_Dimensions)

```
Note: It works but takes one cell to copy at one time. How can I fasten the write process?

**Comment by MariusWirtz** [comment]
Check out this help article. It explains how to write to TM1 most efficiently with TM1py:

https://cubewise-code.github.io/tm1py-tales/2021/write-with-tm1py.html

**Reply by ishmeets920** [reply]
Hi @MariusWirtz,
Thanks for replying. 

Needed to know how can I create this type of dataframe dynamically for the sourceview i fetched. 

data = pd.DataFrame({
        'd1': ['e1', 'e2', 'e4'],
        'd2': ['e2', 'e3', 'e1'],
        'value': [11.5, 12.3, 7.5]
    })

Let say if i have more number of dimensions.

**Comment by MariusWirtz** [comment]
Here is a sample with more dimensions. Does that help you?

``` python
data = pd.DataFrame({
'd1': ['e1', 'e2', 'e4'],
'd2': ['e2', 'e3', 'e1'],
'd3': ['e1', 'e4', 'e2'],
'measure': ['value', 'value', 'value'],
'value': [11.5, 12.3, 7.5]
})
```

**Reply by ishmeets920** [reply]
I want to create this thing dynamically based on the length of elements in dimensions. 
Like, if i have total of 4 dimensions like 

Client = [E1, E2, E3, E4, E5]
Server = [F1, F2, F3, F4]
Lines = [A1........A200]
Data = [F1....F100000]

I have to create a dataframe based on what I fetch from cube dynamically, like the length of 
Lines and values in the Data dimension for lines.

**Reply by MariusWirtz** [reply]
You don't need to use data frame at all if you don't want to. You can use the `write` function which takes a dictionary as input. You can read about it here: https://cubewise-code.github.io/tm1py-tales/2021/write-with-tm1py.html

Whatever structure you use, it must map to the cube!
So if you have a cube with 5 dimensions, your data frame must have 6 columns (5 dimensions + value column).
If you use a dictionary, your keys will be tuples of 5 elements.

**Reply by ishmeets920** [reply]
I have tried with write function with this code 

cellset = {(Dimensionlist[0], Dimensionlist[1], Dimensionlist[2], Dimensionlist[3]): Value for
            Dimensionlist[0], Dimensionlist[1], Dimensionlist[2], Dimensionlist[3], Value in df.values}
 
 tm2.cubes.cells.write_values(source_cube, cellset,skip_non_updateable=True, dimensions=Target_Dimensions)

It is taking one cell and then copying it and so on.
I want to copy a complete line let say 10 cells in a row at single time. Is that possible ?

**Reply by MariusWirtz** [reply]
Please read the help article carefully and you will understand how to work with the different `write` functions. It is all explained and samples are provided for each function.
The write functions do **NOT** write "one cell at a time". If they do, you are doing something wrong in your code.
